functions/lessonNotifications.js

```diff
@@ -94,6 +94,113 @@ module.exports = {
     }
   },
 
+  notifyStudentInstantLessonConfirmed: async (lesson) => {
+        const student = String(lesson.studentUid).split(",")[0];
+
+        try {
+          // Fetch the tutor's profile
+          const studentQuery = await db.collection("profiles").where("uid", "==", studentUid).limit(1).get();
+
+          if (studentQuery.empty) {
+            console.error("Tutor profile not found.");
+            return;
+          }
+
+          const studentData = studentQuery.docs[0].data();
+
+          // Prepare the notification payload for the tutor
+          const studentPayload = preparePayload(
+            "Instant Lesson Confirmed!",
+            `Great news, a tutor is available to help you now!`,
+            studentData.token
+          );
+
+          // Send the notification
+          await sendNotification(studentPayload);
+          console.log("Notification sent to student for lesson confirmation.");
+        } catch (error) {
+          console.error("Error in notifyStudentInstantLessonConfirmed:", error);
+        }
+      },
+
+    notifyTutorCancelled: async (lesson) => {
+      const studentUid = String(lesson.studentUid).split(",")[0];
+
+      try {
+        const studentDoc = await db.collection("profiles").where("uid", "==", studentUid).limit(1).get();
+
+         if (studentDoc.empty) {
+            console.error("Tutor profile not found.");
+            return;
+         }
+
+         const studentData = studentDoc.docs[0].data();
+         const studentPayload = preparePayload(
+            "Your lesson as been cancelled.",
+            `A tutor has cancelled his lesson with you.`,
+            tutorData.token
+         );
+
+         // Send notification
+         await sendNotification(tutorPayload);
+      } catch (error) {
+        console.error("Error in notifyStudentCancelled:", error);
+      }
+
+    },
+
+    // Notify the tutor when a student cancels a lesson he offered to teach
+    notifyStudentCancelled: async (lesson) => {
+       const tutorUid = String(lesson.tutorUid).split(",")[0];
+
+       try {
+         const tutorDoc = await db.collection("profiles").where("uid", "==", tutorUid).limit(1).get();
+
+          if (tutorDoc.empty) {
+            console.error("Tutor profile not found.");
```

Comment:
Good consistency with the structure in these similar code blocks but you might consider making a function that takes a role with the Doc and the notification message and then does the same process everytime. As you might see the similarity of it.

Reply:
Be careful with the description of your PR, I think that you can still imrpove it. Show some pictures of the notifications would be nice and you mentioned that getProfiles was changed which is not really the case. But you do not mention all the .js files that you change. This helps the reviewer understand some context and the big lines of the PR. But keep up the good work, you're improvements are impressive.